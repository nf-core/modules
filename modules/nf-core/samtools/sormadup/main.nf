process SAMTOOLS_SORMADUP {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/samtools:1.19.2--h50ea8bc_0' :
        'biocontainers/samtools:1.19.2--h50ea8bc_0' }"

    input:
    tuple val(meta), path(input)
    tuple val(meta2), path(fasta)

    output:
    tuple val(meta), path("*.bam")      , emit: bam,  optional: true
    tuple val(meta), path("*.cram")     , emit: cram, optional: true
    tuple val(meta), path("*.csi")      , emit: csi,  optional: true
    tuple val(meta), path("*.crai")     , emit: crai, optional: true
    tuple val(meta), path("*.metrics")  , emit: metrics
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def args4 = task.ext.args4 ?: ''
    def args5 = task.ext.args5 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extension = args5.contains("--output-fmt sam") ? "sam" :
                    args5.contains("--output-fmt cram") ? "cram" :
                    "bam"
    def reference = fasta ? "--reference ${fasta}" : ""
    def sort_memory = (task.memory.mega/task.cpus*0.75).intValue()

    """
    samtools cat \\
        --threads $task.cpus \\
        $args \\
        ${input}  \\
    | \\
    samtools collate \\
        $args2 \\
        -O \\
        -u \\
        -T ${prefix}.collate \\
        --threads $task.cpus \\
        ${reference} \\
        - \\
    | \\
    samtools fixmate \\
        $args3 \\
        -m \\
        -u \\
        --threads $task.cpus \\
        - \\
        - \\
    | \\
    samtools sort \\
        $args4 \\
        -u \\
        -T ${prefix}.sort \\
        --threads $task.cpus \\
        -m ${sort_memory}M \\
        - \\
    | \\
    samtools markdup \\
        -T ${prefix} \\
        -f ${prefix}.metrics \\
        --threads $task.cpus \\
        $args5 \\
        - \\
        ${prefix}.${extension}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def args5 = task.ext.args5 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extension = args5.contains("--output-fmt sam") ? "sam" :
                    args5.contains("--output-fmt cram") ? "cram" :
                    "bam"

    """
    touch ${prefix}.${extension}
    touch ${prefix}.metrics

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
