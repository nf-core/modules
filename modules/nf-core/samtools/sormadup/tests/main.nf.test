nextflow_process {

    name "Test Process SAMTOOLS_SORMADUP"
    script "../main.nf"
    process "SAMTOOLS_SORMADUP"

    tag "modules"
    tag "modules_nfcore"
    tag "samtools"
    tag "samtools/sormadup"

    test("sarscov2 - bam") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'fasta' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("sarscov2 - bam - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'fasta' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
