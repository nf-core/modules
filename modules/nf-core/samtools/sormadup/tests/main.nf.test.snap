{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.metrics:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,cd361852bb57be36b152d408121eab50"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "crai": [
                    
                ],
                "cram": [
                    
                ],
                "csi": [
                    
                ],
                "metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.metrics:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cd361852bb57be36b152d408121eab50"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-05T07:47:32.119697"
    },
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.bam:md5,33b4c8ab46856f9ced3a9f86628f268c"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.metrics:md5,bed5fec5c6c55500f10042eba21f9d13"
                    ]
                ],
                "5": [
                    "versions.yml:md5,cd361852bb57be36b152d408121eab50"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.bam:md5,33b4c8ab46856f9ced3a9f86628f268c"
                    ]
                ],
                "crai": [
                    
                ],
                "cram": [
                    
                ],
                "csi": [
                    
                ],
                "metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.metrics:md5,bed5fec5c6c55500f10042eba21f9d13"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cd361852bb57be36b152d408121eab50"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-05T07:47:09.878487"
    }
}